app/gateway/2.8.x/admin-api/developers/reference.md

```diff
@@ -0,0 +1,1417 @@
+
+
+**Request Body**
+
+Attribute                     | Description
+---------:                    | --------   
+`meta`                        | Metadata for the account in JSON format. Accepts fields defined in the Dev Portal settings. <br><br> By default, the meta attribute requires a `full_name` field. You can remove this requirement, or add other fields as neccessary. <br><br> For example: `meta: {"full_name":"<NAME>"}`.
+`email`       | The email of the developer to create. This becomes their login username.
+
+
+**Response**
+
+```
+HTTP/1.1 200 OK
+```
+
+```json
+{
+    "developer": {
+        "consumer": {
+            "id": "855fe098-3bb3-4041-8a3d-149a9e5462c6"
+        },
+        "created_at": 1644617754,
+        "email": "example@example.com",
+        "id": "f94e4fbd-c30b-41d7-b349-13c8f4fc23ca",
+        "rbac_user": {
+            "id": "7d40540b-4b65-4fa6-b04e-ffe6c2c40b0d"
+        },
+        "roles": [
+            "QA"
+        ],
+        "status": 0,
+        "updated_at": 1644621148
+    }
+}
+```
+
+### Delete a developer
+
+**Endpoint**
+
+<div class="endpoint delete">/developers/{DEVELOPER_ID}</div>
+
+Attribute                     | Description
+---------:                    | --------   
+`{DEVELOPER_ID}`  | The UUID of the developer you want to delete.
+
+**Response**
+
+```
+HTTP/1.1 204 No Content
+```
+
+## Export developer metadata
+
+Export a list of developers and their metadata into a CSV file.
+
+Creates a file named `developers.csv` in the present working directory.
+
+<div class="endpoint get">/developers/export</div>
+
+**Response**
+
+```
+HTTP/1.1 200 OK
+```
+
+```json
+something
+```
+
+## Roles
+
+### List all Dev Portal roles
+
+**Endpoint**
+
+<div class="endpoint get">/developers/roles</div>
+
+**Response**
+
+```
+HTTP/1.1 200 OK
+```
+
+```json
+{
+    "data": [
+        {
+            "created_at": 1644615388,
+            "id": "525694de-776c-43ad-a940-769295d0c54e",
+            "name": "Application Developer"
+        },
+        {
+            "created_at": 1644619493,
+            "id": "c8535760-e37f-4578-921f-04e781a3f8f1",
+            "name": "QA"
+        }
+    ],
+    "next": null,
+    "total": 2
+}
+```
+
+### Create a Dev Portal role
+
+When you create a role, by default, the role has access to all content in Dev Portal.
+
+**Endpoint**
+
+<div class="endpoint post">/developers/roles</div>
+
+**Request body**
+
+Attribute                   | Description
+---------:                  | --------   
+`name` <br>*required*       | A name for the role.
+`comment`                   | A description of the role.
+`permissions`               | [_Note: I couldn't figure out how to make this work_] Permissions to assign to the role, in JSON format. If you don't provide any specific permissions, the role has access to all applications and specs by default. <br><br> Example: `permissions={"specs/petstore.yaml"}`
```

Comment:
Couldn't figure out the permissions format, kept getting errors saying that it doesn't know what `permissions` is:

```
{
    "code": 2,
    "fields": {
        "permissions": "unknown field"
    },
    "message": "schema violation (permissions: unknown field)",
    "name": "schema violation"
}
```
I can't entirely tell if I just have the wrong format, or if you can't set permissions through the Admin API (which doesn't seem right, since you can do it through the GUI and there's a `permissions` field in the response).

Reply:
The permissions field is read-only in the Admin API for roles. Permissions are managed through the headmatter of the portal pages and parsed out to generate the permissions that are returned when retrieving a role.

https://docs.konghq.com/gateway/2.8.x/developer-portal/administration/developer-permissions/#readable_by-attribute

To update the permissions for a role through the API, you have to make a request to files API like:
```
curl 'http://localhost:8001/default/files/content/dashboard.txt' -X PATCH \
  --data-raw '{"contents":"---\nreadable_by:\n  - role1\nlayout: system/dashboard.html\nprivate: true\ntitle: Dashboard\n---\n\n"}'
```

But doing this requires you to first GET the file, parse out the headmatter, update the `readable_by` roles array, stringify it back into the file, and then PATCH the file back. I don't think it is recommended to do via API, but rather by using the KM settings, updating the files in the KM portal files editor, or through the portal CLI.